From 4735bd07a7227f54c84be4bc5191ac79ce2524c7 Mon Sep 17 00:00:00 2001
From: Federico Mena Quintero <federico@gnome.org>
Date: Wed, 6 Nov 2019 18:03:27 -0600
Subject: [PATCH] (#525): Consider specificity when applying CSS selector
 matches

We still don't consider ordering of rules in stylesheets, but at least
we handle specificity for conflicting matches with this.

Fixes https://gitlab.gnome.org/GNOME/librsvg/issues/525
---
 rsvg_internals/src/css.rs                     |   8 ++++--
 1 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/rsvg_internals/src/css.rs b/rsvg_internals/src/css.rs
index c8ae072c..dee9e9ba 100644
--- a/rsvg_internals/src/css.rs
+++ b/rsvg_internals/src/css.rs
@@ -282,7 +282,7 @@ impl CssRules {
     }
 
     pub fn get_matches(&self, node_data: &NodeData) -> Vec<Selector> {
-        self.selectors_to_declarations
+        let mut matches: Vec<_> = self.selectors_to_declarations
             .iter()
             .filter_map(|(selector, _)| {
                 if self.selector_matches_node(selector, node_data) {
@@ -292,7 +292,11 @@ impl CssRules {
                 }
             })
             .map(Selector::clone)
-            .collect()
+            .collect();
+
+        matches.as_mut_slice().sort_by(|sel_a, sel_b| sel_a.specificity.cmp(&sel_b.specificity));
+
+        matches
     }
 }

-- 
2.22.0

